network/src/state.rs

```diff
@@ -0,0 +1,173 @@
+// Copyright (C) 2019-2022 Aleo Systems Inc.
+// This file is part of the snarkOS library.
+
+// The snarkOS library is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// The snarkOS library is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with the snarkOS library. If not, see <https://www.gnu.org/licenses/>.
+
+use std::{net::SocketAddr, sync::Arc};
+
+use once_cell::race::OnceBox;
+use snarkos_environment::{helpers::NodeType, Environment};
+use snarkvm::prelude::*;
+use tokio::sync::oneshot;
+
+use crate::{
+    ledger::{Ledger, LedgerHandler},
+    operator::{Operator, OperatorHandler},
+    peers::{Peers, PeersHandler},
+    prover::{Prover, ProverHandler},
+};
+
+pub struct State<N: Network, E: Environment> {
+    /// The local address of the node.
+    pub local_ip: SocketAddr,
+    /// The list of peers for the node.
+    peers: OnceBox<Peers<N, E>>,
+    /// The ledger of the node.
+    ledger: OnceBox<Ledger<N, E>>,
+    /// The prover of the node.
+    prover: OnceBox<Prover<N, E>>,
+    /// The operator of the node.
+    operator: OnceBox<Operator<N, E>>,
+}
+
+impl<N: Network, E: Environment> State<N, E> {
+    pub fn new(local_ip: SocketAddr) -> Self {
+        Self {
+            local_ip,
+            peers: Default::default(),
+            ledger: Default::default(),
+            operator: Default::default(),
+            prover: Default::default(),
+        }
+    }
+
+    pub async fn initialize_peers(self: &Arc<Self>, peers: Peers<N, E>, mut peers_handler: PeersHandler<N, E>) {
```

Comment:
it's because they are called by the `Server`, and can also be called in tests that don't require the functionalities of the `Server` (i.e. just the state)